zk-sdk/src/transcript.rs

```diff
@@ -0,0 +1,143 @@
+use {
+    crate::errors::TranscriptError,
+    curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar, traits::IsIdentity},
+    merlin::Transcript,
+};
+
+pub trait TranscriptProtocol {
+    /// Append a domain separator for an `n`-bit rangeproof for ElGamalKeypair
+    /// ciphertext using a decryption key
+    fn rangeproof_from_key_domain_separator(&mut self, n: u64);
+
+    /// Append a domain separator for an `n`-bit rangeproof for ElGamalKeypair
+    /// ciphertext using an opening
+    fn rangeproof_from_opening_domain_separator(&mut self, n: u64);
+
+    /// Append a domain separator for a length-`n` inner product proof.
+    fn innerproduct_domain_separator(&mut self, n: u64);
+
+    /// Append a domain separator for close account proof.
+    fn close_account_proof_domain_separator(&mut self);
+
+    /// Append a domain separator for withdraw proof.
+    fn withdraw_proof_domain_separator(&mut self);
+
+    /// Append a domain separator for transfer proof.
+    fn transfer_proof_domain_separator(&mut self);
+
+    /// Append a `scalar` with the given `label`.
+    fn append_scalar(&mut self, label: &'static [u8], scalar: &Scalar);
+
+    /// Append a `point` with the given `label`.
+    fn append_point(&mut self, label: &'static [u8], point: &CompressedRistretto);
+
+    /// Append a domain separator for equality proof.
+    fn equality_proof_domain_separator(&mut self);
+
+    /// Append a domain separator for zero-balance proof.
+    fn zero_balance_proof_domain_separator(&mut self);
+
+    /// Append a domain separator for grouped ciphertext validity proof.
+    fn grouped_ciphertext_validity_proof_domain_separator(&mut self);
+
+    /// Append a domain separator for batched grouped ciphertext validity proof.
+    fn batched_grouped_ciphertext_validity_proof_domain_separator(&mut self);
+
+    /// Append a domain separator for fee sigma proof.
+    fn fee_sigma_proof_domain_separator(&mut self);
+
+    /// Append a domain separator for public-key proof.
+    fn pubkey_proof_domain_separator(&mut self);
+
+    /// Check that a point is not the identity, then append it to the
+    /// transcript.  Otherwise, return an error.
+    fn validate_and_append_point(
+        &mut self,
+        label: &'static [u8],
+        point: &CompressedRistretto,
+    ) -> Result<(), TranscriptError>;
+
+    /// Compute a `label`ed challenge variable.
+    fn challenge_scalar(&mut self, label: &'static [u8]) -> Scalar;
+}
+
+impl TranscriptProtocol for Transcript {
+    fn rangeproof_from_key_domain_separator(&mut self, n: u64) {
+        self.append_message(b"dom-sep", b"rangeproof from opening v1");
+        self.append_u64(b"n", n);
+    }
+
+    fn rangeproof_from_opening_domain_separator(&mut self, n: u64) {
+        self.append_message(b"dom-sep", b"rangeproof from opening v1");
+        self.append_u64(b"n", n);
+    }
```

Comment:
Are these meant to be the same? It seems like the first one should be a different name

Reply:
Yeah so the `rangeproof_from_key_...` should append the message `rangeproof from key v1`. We also seem to have function `close_account_proof_domain_separator`, which is not deprecated and not used any more... I will clean up transcript on a follow-up. Sorry I should have caught this when I am migrating things over.